### PR TITLE
Efficiently Pack Uint64 Lists

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Cleanup

**What does this PR do? Why is it needed?**

Currently we inefficiently pack in uint64 lists into its respective ssz chunks. We allocate more memory than is required to pack them in. This PR introduces a new method which does it in one step and handles large lists more efficiently.

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/state/stateutil/validator_root.go [hunk 1]
```diff
@@ -87,3 +83,40 @@ func Uint64ListRootWithRegistryLimit(balances []uint64) ([32]byte, error) {
 	binary.LittleEndian.PutUint64(balancesLengthRoot, uint64(len(balances)))
 	return ssz.MixInLength(balancesRootsRoot, balancesLengthRoot), nil
 }
+
+// PackUint64IntoChunks packs a list of uint64 values into 32 byte roots.
+func PackUint64IntoChunks(vals []uint64) ([][32]byte, error) {
```
Comment (rkapka) [inline beacon-chain/state/stateutil/validator_root.go:88]: Shouldn't this be called `PackBalancesIntoChunks`? You use balances in code and comments inside the function.

File: beacon-chain/state/stateutil/validator_root.go [hunk 2]
```diff
@@ -87,3 +83,40 @@ func Uint64ListRootWithRegistryLimit(balances []uint64) ([32]byte, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	// Determine total number of chunks to be
+	// allocated to provided list of unsigned
+	// 64-bit integers.
+	numOfChunks := len(vals) / iNumOfElems
+	if len(vals)%iNumOfElems != 0 {
+		numOfChunks++
+	}
```
Comment (rkapka) [inline beacon-chain/state/stateutil/validator_root.go:105]: Can you add a comment explaining what this does?

File: beacon-chain/state/stateutil/validator_root.go [hunk 3]
```diff
@@ -87,3 +83,40 @@ func Uint64ListRootWithRegistryLimit(balances []uint64) ([32]byte, error) {
...
+		// In order to determine how to pack in the uint64 value by index into
+		// our balance chunk list we need to determine a few things.
+		// 1) The chunk which the particular uint64 value corresponds to.
+		// 2) The position of the value in the chunk itself.
+		//
+		// Once we have determined these 2 values we can simply find the correct
+		// section of contiguous bytes to insert the value in the chunk.
+		chunkIdx := idx / iNumOfElems
+		idxInChunk := idx % iNumOfElems
+		sizeOfElem := 32 / iNumOfElems
```
Comment (rkapka) [inline beacon-chain/state/stateutil/validator_root.go:117]: This doesn't have to be recalculated in each iteration.

File: beacon-chain/state/stateutil/validator_root.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/state/stateutil/validator_root.go:88]: This is applied for all uint64 lists not just for balances.

**Review by rkapka — APPROVED** [review]

**Pull request merged**